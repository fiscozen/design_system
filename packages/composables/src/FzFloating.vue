<script lang="ts" setup>
import { ref, useSlots, watch, toRef, computed } from 'vue'
import { useFloating } from './composables'
import { FzFloatingProps, FzUseFloatingArgs } from './types'

const props = withDefaults(defineProps<FzFloatingProps>(), {
  position: 'auto',
  isOpen: false
})

const opener = ref(null)
const content = ref<HTMLElement|null>(null)

const slots = useSlots()

const useFloatingOpts: FzUseFloatingArgs = {
  position: computed(() => props.position),
  element: {
    domRef: content
  },
  container: {
    domRef: toRef(props.container || document.body)
  }
}
if (slots.opener) {
  useFloatingOpts.opener = {
    domRef: opener
  }
}

const floating = useFloating(useFloatingOpts)

watch(
  () => props.position,
  () => props.isOpen && floating.setPosition()
)
watch(
  () => props.isOpen,
  (newVal) => {
    if (!newVal || !content.value) {
      return;
    }
    content.value.style.top = '0px';
    content.value.style.left = '0px';
    content.value.style.transform = 'none';
    floating.setPosition()
  }
)
</script>

<template>
  <div>
    <div ref="opener" class="inline-flex">
      <slot name="opener" :isOpen :floating></slot>
    </div>
    <div
      ref="content"
      v-show="!$slots.opener || ($slots.opener && isOpen)"
      class="bg-core-white absolute p-4"
      :class="contentClass"
    >
      <slot :isOpen :floating></slot>
    </div>
  </div>
</template>

<style></style>
