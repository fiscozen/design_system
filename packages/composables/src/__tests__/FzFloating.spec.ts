import { describe, it, expect } from 'vitest'

import { mount } from '@vue/test-utils'
import FzFloating from '../FzFloating.vue'

describe('FzFloating', () => {
  it('should match snapshot', () => {
    const wrapper = mount(FzFloating, {
    })

    expect(wrapper.html()).toMatchSnapshot()
  })
})
