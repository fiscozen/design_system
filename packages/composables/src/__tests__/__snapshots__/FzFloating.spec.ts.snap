// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`FzFloating > should match snapshot 1`] = `
"<div>
  <div class="inline-flex"><button>opener</button></div>
  <div class="bg-core-white absolute p-4 z-10">
    <div>content</div>
  </div>
</div>"
`;

exports[`FzFloating > should match snapshot 2`] = `
"<div>
  <div class="inline-flex"><button>opener</button></div>
  <div class="bg-core-white absolute p-4 z-10">
    <div>content</div>
  </div>
</div>"
`;

exports[`FzFloating > should match snapshot 3`] = `
"<div>
  <div class="inline-flex"><button>opener</button></div>
  <div class="bg-core-white absolute p-4 z-10">
    <div>content</div>
  </div>
</div>"
`;

exports[`FzFloating > should match snapshot 4`] = `
"<div>
  <div class="inline-flex"><button>opener</button></div>
  <div class="bg-core-white absolute p-4 z-10">
    <div>content</div>
  </div>
</div>"
`;

exports[`FzFloating > should match snapshot 5`] = `
"<div>
  <div class="inline-flex"><button>opener</button></div>
  <div class="bg-core-white absolute p-4 z-10">
    <div>content</div>
  </div>
</div>"
`;

exports[`FzFloating > should match snapshot 6`] = `
"<div>
  <div class="inline-flex"><button>opener</button></div>
  <div class="bg-core-white absolute p-4 z-10">
    <div>content</div>
  </div>
</div>"
`;

exports[`FzFloating > should match snapshot 7`] = `
"<div>
  <div class="inline-flex"><button>opener</button></div>
  <div class="bg-core-white absolute p-4 z-10">
    <div>content</div>
  </div>
</div>"
`;

exports[`FzFloating > should match snapshot 8`] = `
"<div>
  <div class="inline-flex"><button>opener</button></div>
  <div class="bg-core-white absolute p-4 z-10">
    <div>content</div>
  </div>
</div>"
`;

exports[`FzFloating > should match snapshot 9`] = `
"<div>
  <div class="inline-flex"><button>opener</button></div>
  <div class="bg-core-white absolute p-4 z-10">
    <div>content</div>
  </div>
</div>"
`;

exports[`FzFloating > should match snapshot 10`] = `
"<div>
  <div class="inline-flex"><button>opener</button></div>
  <div class="bg-core-white absolute p-4 z-10">
    <div>content</div>
  </div>
</div>"
`;

exports[`FzFloating > should match snapshot 11`] = `
"<div>
  <div class="inline-flex"><button>opener</button></div>
  <div class="bg-core-white absolute p-4 z-10">
    <div>content</div>
  </div>
</div>"
`;

exports[`FzFloating > should match snapshot 12`] = `
"<div>
  <div class="inline-flex"><button>opener</button></div>
  <div class="bg-core-white absolute p-4 z-10">
    <div>content</div>
  </div>
</div>"
`;
